website/blog/2024-03-18-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,230 @@
+
+The diagram below is a visualization of how finetuning and MLflow fit together.
+
+![](finetune_with_MLflow.png)
+
+Using MLflow while fine tuning ML model in Databricks
+
+
+# The Importance of Logging
+
+The integration of the Hugging Face trainer callback with MLFlow plays a pivotal role in this ecosystem. It allows for comprehensive tracking and logging of metrics, parameters, and artifacts during model finetuning. This logging can be configured flexibly – at every epoch, after a set number of epochs, or steps. Importantly, it can cover the entire model checkpoint or be limited to just the adapter when using [PEFT (Parameter-Efficient Fine-Tuning)](https://huggingface.co/docs/peft/index) methods like [QLoRA](https://arxiv.org/abs/2305.14314). We have highlighted the additional code necessary for MLflow to log every epoch and checkpoint.
+
+
+## Configuring Hugging Face Trainer Callback for MLflow
+
+```python
+class CustomMLflowCallback(MLflowCallback):
+    """
+    A [`TrainerCallback`] that sends the logs to [MLflow](https://www.mlflow.org/). Can be disabled by setting
+    environment variable `DISABLE_MLFLOW_INTEGRATION = TRUE`.
+    """
+
+  def on_save(self, args, state, control,model,tokenizer, **kwargs):
+        if self._initialized and state.is_world_process_zero and self._log_artifacts:
+           ckpt_dir = f"checkpoint-{state.global_step}"
+  	 self._ml_flow.transformers.log_model(
+                             transformers_model={
+                                 "model": model,
+                                 "tokenizer": tokenizer,
+                             },
+                             task="text-generation",
+                             artifact_path=ckpt_dir,
+                             pip_requirements=["torch", "transformers", "accelerate"],
+                             input_example=input_example,
+                             signature=signature,
+                             metadata={"task": "llm/v1/completions"}
+                         )
+```
+
+See detailed code block [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/495c156f73435996706c42bfda4bb7519dc21dab/databricks_llm/callbacks.py#L64)
+
+In the code snippet above, we create a new callback class that includes an additional method, **on_save**, which logs the weights and tokenizer after **every n** steps/epochs. This frequency is configured as part of the training arguments using the Transformer library. Additionally, we can employ a custom pyfunc to perform the same activity. For an example, please refer to the link provided [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/main/databricks_llm/callbacks.py).
+
+The custom callback gracefully saves the model checkpoints during the finetuning process. The mlflow\.transformers.log\_model expression saves the model and the tokenizer alongside the dependencies, input example and the model signature during each logging event.
```

Comment:
This saves both PEFT weights, the component elements, and the foundation model weights with each checkpoint? 
This is definitely not something that we would want to show users as it would make a static copy of the foundation model for each checkpoint (and the approach of PEFT doesn't modify the foundation model at all), causing extreme file duplication and a massive slowdown to the training process as it serializes massive files with each n training iterations. 
While saving the PEFT weights might make sense (I'm not familiar with how long it takes to use PEFT training with this model), do we want to show users a customized implementation of checkpointing for this library? 
I think that showing the happy path of the new features introduced in 2.11 is more relevant than showing something like this, particularly since this will likely increase the training time to the point that a user would feel very frustrated at the sheer volume of copied data being generated and how long it would take to run (which would directly cause them to assume that MLflow is terrible for PEFT).